bittide/src/Clash/Cores/Xilinx/SystemMonitor.hs

```diff
@@ -0,0 +1,266 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE NoFieldSelectors #-}
+
+module Clash.Cores.Xilinx.SystemMonitor (
+  Status (..),
+  temperatureMonitorCelcius,
+  temperatureMonitor,
+  adcToTemperature,
+) where
+
+import Clash.Prelude
+
+import Clash.Cores.Xilinx.Xpm.Cdc.Internal
+import Clash.Functor.Extra ((<<$>>))
+
+{- | A wrapper for a SYSMONE1 instance which only monitors the FPGA temperature. Converts the ADC
+value of the temperature sensor to the temperature in degree Celcius.
+
+XXX: This function can lead to timing issues because the instantiated DSP does
+     not include input/output registers.
+-}
+temperatureMonitorCelcius ::
+  forall dom.
+  (HiddenClockResetEnable dom, 1 <= DomainPeriod dom) =>
+  Signal dom (Status, Maybe (Signed 11))
+temperatureMonitorCelcius = bundle (status, temperature)
+ where
+  (status, adc) = unbundle temperatureMonitor
+  temperature = toSigned <<$>> (dflipflop $ adcToTemperature <<$>> (dflipflop adc))
+   where
+    toSigned ::
+      forall int frac. (KnownNat int, KnownNat frac) => SFixed int frac -> Signed int
+    toSigned = (resize . bitCoerce) . (flip shiftR (natToNum @frac))
+
+{- | Calculates the temperature based on the ADC value. Assumes the temperature sensor uses the
+on-chip reference. For more information see:
+
+    https://docs.amd.com/v/u/en-US/ug580-ultrascale-sysmon
+-}
+adcToTemperature ::
+  forall adcDepth.
+  (KnownNat adcDepth) =>
+  BitVector adcDepth ->
+  SFixed (adcDepth + 1) 14
+adcToTemperature adc = t
+ where
+  t = adcSF * (a `shiftR` (natToNum @adcDepth)) + b
+  adcSF :: SFixed (adcDepth + 1) 14
+  adcSF = bitCoerce ((resize adc) `shiftL` 14)
+  a = 501.3743
+  b = -273.6777
+
+{- | A wrapper for a SYSMONE1 instance which only monitors the FPGA temperature and outputs the
+ADC value.
+-}
+temperatureMonitor ::
+  forall dom.
+  (HiddenClockResetEnable dom, 1 <= DomainPeriod dom) =>
+  Signal dom (Status, Maybe (BitVector 10))
+temperatureMonitor = bundle (status, tOut)
+ where
+  tOut = mux dRdy (Just <$> adc) (pure Nothing)
+
+  -- The 10-bit ADC value is stored in the MSBs.
+  adc = resize . (flip shiftR 6) <$> dO
+  (status, dRdy, dO) = unbundle $ sysMon @dom hasClock hasReset dEn dAddr dWe dI
+
+  dEn = (counter .==. (pure 0))
+  dWe = pure False
+  dAddr = pure 0x00
+  dI = pure 0x00
+
+  counter = register (0 :: Index 10) $ satSucc SatWrap <$> counter
+
+data Status = Status
+  { busy :: Bool
+  , channel :: BitVector 6
+  , endOfConversion :: Bool
+  , endOfSequence :: Bool
+  }
+  deriving (Eq, Generic, NFDataX)
+
+-- The minimum period of the ADC clock associated with the maximum frequency of 5.2 MHz
+type MinAdcPeriod = 192_308
+type AdcClockDiv dom = Max 2 (MinAdcPeriod `DivRU` (DomainPeriod dom))
```

Comment:
Where does the 5.2MHz come from?

I guess this is a period in Picoseconds, could you use some type synonyms to make this more obvious? https://github.com/clash-lang/clash-compiler/blob/master/clash-prelude/src/Clash/Signal/Internal.hs#L455-L486

Reply:
Does this clarify it for you?

```haskell
{- The internal ADC clock is derived from the DRP clock. The SYSMON supports a DRP clock
frequency up to 250 MHz, whereas the The maximum frequency of the ADC clock is 5.2 MHz.
A clock divider with a minimal value of 2 is used to get the appropriate ADC clock
frequency.
-}
type AdcClockDiv dom = Max 2 (HzToPeriod 5_200_000 `DivRU` DomainPeriod dom)
```

Reply:
That's fine :)